### PR TITLE
perf: Speed up rule deduplication

### PR DESCRIPTION
**Pull request by underyx**
PR checklist:

- [ ] Documentation is up-to-date
- [ ] Changelog is up-to-date
- [ ] Change has no security implications (otherwise, ping security team)

### CODE REVIEW

File: semgrep/semgrep/rule.py [hunk 1]
```diff
@@ -247,6 +246,7 @@ def has_runnable_rule(d: Dict[str, Any]) -> bool:
 
 
 def rule_without_metadata(rule: Rule) -> Rule:
-    rule2 = deepcopy(rule)
```
Comment (nbrahms) [inline semgrep/semgrep/rule.py:250]: what are the implications of now doing a shallow copy?

Comment (underyx) [reply semgrep/semgrep/rule.py:250]: Nothing! We only ever use this to create a read-only dictionary key.

Comment (underyx) [reply semgrep/semgrep/rule.py:250]: I'd like to make `Rule` a frozen attrs class to make sure it's immutable. But that's much more refactoring work.

**Review by ievans — APPROVED** [review]

**Comment by emjin** [comment]
Oops, this was me. Thanks for fixing it!

**Pull request merged**